File: test/corpus/propagation/let_comparison_propagation.exs
defmodule RustQ.Corpus.Propagation.LetComparisonPropagation do
  @moduledoc "Propagate a fallible let RHS when downstream comparisons use the success value."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @type field :: %{required(:id) => R.u32()}

  @spec read_id() :: R.nif_result(R.u32())
  defrust(read_id(), do: {:ok, 0})

  @spec run(R.slice(field())) :: R.nif_result(R.unit())
  defrust run(fields) do
    field_id = read_id()

    if field_id == 0 do
      :ok
    else
      case fields.binary_search_by_key(ref(field_id), fn field -> field.id end) do
        {:ok, _index} -> :ok
        {:error, _index} -> {:error, badarg()}
      end
    end
  end
end

File: test/corpus/propagation/let_comparison_propagation.rs
#[derive(Clone, Debug)]
pub struct Field {
    pub id: u32,
}

pub fn decode_field<'a>(term: Term<'a>) -> NifResult<Field> {
    Ok(Field {
        id: term.map_get(atoms::id())?.decode()?,
    })
}

fn read_id() -> NifResult<u32> {
    Ok(0)
}

fn run(fields: &[Field]) -> NifResult<()> {
    let field_id = read_id()?;
    if field_id == 0 {
        Ok(())
    } else {
        match fields.binary_search_by_key(&field_id, |field| field.id) {
            Ok(_index) => Ok(()),
            Err(_index) => Err(rustler::Error::BadArg),
        }
    }
}
